app/store/middleware.js

```diff
@@ -0,0 +1,201 @@
+        payload: nextState,
+        error: action.error
+    };
+}
+
+function getLastChannelForTeam(payload) {
+    const lastChannelForTeam = {...payload.views.team.lastChannelForTeam};
+    const convertLastChannelForTeam = Object.values(lastChannelForTeam).some((value) => !Array.isArray(value));
+
+    if (convertLastChannelForTeam) {
+        Object.keys(lastChannelForTeam).forEach((id) => {
+            lastChannelForTeam[id] = [lastChannelForTeam[id]];
+        });
+    }
+
+    return lastChannelForTeam;
+}
+
+function cleanupState(action, keepCurrent = false) {
+    const {payload: resetPayload} = resetStateForNewVersion(action);
+    const {payload} = action;
+    const {currentChannelId} = payload.entities.channels;
+    const {statuses, ...otherUsers} = payload.entities.users; //eslint-disable-line no-unused-vars
+
+    const {lastChannelForTeam} = resetPayload.views.team;
+    const nextEntitites = {
+        posts: {
+            posts: {},
+            postsInChannel: {},
+            reactions: {},
+            selectedPostId: payload.entities.posts.selectedPostId,
+            currentFocusedPostId: payload.entities.posts.currentFocusedPostId
+        },
+        files: {
+            files: {},
+            fileIdsByPostId: {}
+        }
+    };
+
+    const retentionPeriod = Config.EnableMessageRetention ? Config.MessageRetentionPeriod + 1 : 0;
+    const postIdsToKeep = Object.values(lastChannelForTeam).reduce((array, channelIds) => {
+        const ids = channelIds.reduce((result, id) => {
+            // we need to check that the channel id is not already included
+            // the reason it can be included is cause at least one of the last channels viewed
+            // in a team can be a DM or GM and the id can be duplicate
+            if (!nextEntitites.posts.postsInChannel[id]) {
+                let postIds;
+                if (keepCurrent && currentChannelId === id) {
+                    postIds = payload.entities.posts.postsInChannel[id];
+                } else {
+                    postIds = payload.entities.posts.postsInChannel[id].slice(0, 60);
+                }
+                nextEntitites.posts.postsInChannel[id] = postIds;
+                return result.concat(postIds);
+            }
+
+            return result;
+        }, []);
+        return array.concat(ids);
+    }, []);
+
+    postIdsToKeep.forEach((postId) => {
+        const post = payload.entities.posts.posts[postId];
+        const skip = keepCurrent && currentChannelId === post.channel_id;
+
+        if (!skip && retentionPeriod && (Date.now() - post.create_at) / (1000 * 3600 * 24) > retentionPeriod) {
+            const postsInChannel = nextEntitites.posts.postsInChannel[post.channel_id];
+            const index = postsInChannel.indexOf(postId);
+            if (index !== -1) {
+                postsInChannel.splice(index, 1);
+            }
+            return;
+        }
+
+        nextEntitites.posts.posts[postId] = post;
```

Comment:
Do we need to do anything to save other posts in a given thread?

Reply:
not really cause in the end we are loading the posts for the thread don't we?

Reply:
I don't know if we are still doing that. I thought the method to do that before was called something like `loadPostsForThread`, but I can't find that action in either the mobile or redux repos